### PR TITLE
Missing functionality in scryer

### PR DESCRIPTION
**Pull request by fayeed**
@cmpute This PR includes all the missing functionality that the Scryer needs that I had to add, please let me know if the implementation makes sense to you or needs to be updated.

### CODE REVIEW

File: integer/src/bits.rs [hunk 1]
```diff
@@ -241,6 +241,11 @@ impl IBig {
             Negative => repr.trailing_ones_neg(),
         }
     }
+
+    #[inline]
+    pub fn count_ones(&self) -> usize {
```
Comment (cmpute) [inline integer/src/bits.rs:246]: Why would you need the functionality to count ones on a signed integer? The implementation here doesn't make sense to me as well...

File: integer/src/cmp.rs [hunk 2]
```diff
@@ -121,3 +121,32 @@ impl AbsCmp<IBig> for UBig {
         self.0.as_typed().cmp(&rhs.0.as_sign_typed().1)
     }
 }
+
+macro_rules! impl_partial_ord_for_prim_ints {
```
Comment (cmpute) [inline integer/src/cmp.rs:125]: For comparison between different types, it's designed to use the `num-order` crate right now. Could you please take a look at that crate and use the `num_cmp` and `num_eq` functions provided in that crate?

File: integer/src/convert.rs [hunk 3]
```diff
@@ -259,6 +259,62 @@ impl IBig {
             Inexact(val, diff) => Inexact(sign * val, sign * diff),
         }
     }
+
```
Comment (cmpute) [inline integer/src/convert.rs:262]: The intended way to convert big integers into small integers is through the `TryFrom` APIs, there is no need to create standalone functions. (besides, convert to floats first is very inefficient).

To do a narrowing conversion, you can do: `let small: i32 = big.try_into().unwrap()`

File: integer/src/div_const.rs [hunk 4]
```diff
@@ -638,3 +638,22 @@ mod repr {
         Repr::from_buffer(lhs)
     }
 }
+
```
Comment (cmpute) [inline integer/src/div_const.rs:641]: Same as above, please use the crate `num-order` instead.

File: Cargo.toml [hunk 5]
```diff
@@ -32,7 +32,7 @@ default-members = ["base", "integer", "float", "rational", "macros"]
 # TODO(next): Assign a version name to all crates that haven't reached v1. For crates that already reaches v1, we assume that the update frequency of that crate is low, so it's unnecessary to create a version tag until v2.0 is released
 #             Our strategy for pre v1 dependencies: the bump of default version is not considered a break change, only minor version is required to bump.
 #             Our strategy for v1 dependencies: the bump of default version is considered a break change, the support for new version will be release through a versioned feature flag, and the default version will be bumped in the next major version.
-default = ["std", "num-order"]
+default = ["std", "num-order", "num-traits", "num-traits_v02"]
```
Comment (cmpute) [inline Cargo.toml:35]: Unfortunately I will not accept adding `num-traits` as a default dependency. Please specify the dependency in your own Cargo.toml

File: integer/src/ibig.rs [hunk 6]
```diff
@@ -68,6 +70,23 @@ use crate::{
 pub struct IBig(pub(crate) Repr);
 
 impl IBig {
+    #[inline]
```
Comment (cmpute) [inline integer/src/ibig.rs:73]: Are you aware of the [DivRem](https://docs.rs/dashu-base/latest/dashu_base/ring/trait.DivRem.html) and [DivRemEuclid](https://docs.rs/dashu-base/latest/dashu_base/ring/trait.DivRemEuclid.html) trait provided in the `dashu-base` crate?

File: integer/src/ubig.rs [hunk 7]
```diff
@@ -71,7 +71,7 @@ pub struct UBig(pub(crate) Repr);
 impl UBig {
     /// Get the representation of UBig.
     #[inline]
-    pub(crate) fn repr(&self) -> TypedReprRef<'_> {
+    pub fn repr(&self) -> TypedReprRef<'_> {
```
Comment (cmpute) [inline integer/src/ubig.rs:74]: Where do you need to access this private representation? This type is not intended for outside use.

File: integer/src/ibig.rs [hunk 8]
```diff
@@ -68,6 +70,23 @@ use crate::{
...
+    pub fn div_rem_floor_ref(&self, rhs: &IBig) -> (IBig, IBig) {
+        let div = self.div(rhs);
+        let rem = self.rem(rhs);
+
+        (div, rem)
+    }
+
+    #[inline]
+    pub fn is_odd(&self) -> bool {
```
Comment (cmpute) [inline integer/src/ibig.rs:82]: Is it necessary to have a standalone function for this? You can test it by taking the first bit, which is very fast.

**Review by cmpute — CHANGES_REQUESTED** [review]
In short, I'm afraid that I can't merge these changes directly. However, I would love to use this PR as a thread to discuss how I should change the `dashu` crates to fit your demand.

Most of the changes proposed in this PR are not necessary (such as proposing new functions for conversions and comparison). The only functions that might be necessary are `RBig::is_integer`, `IBig::count_ones`, `IBig::is_odd`. I will add the first one in the next release, but I'm not very satisfied with the other two...

**Comment by cmpute** [comment]
And I apologize that the documentations for `dashu` are pretty bad right now. I do plan to write a guide about how to use these crates, but I haven't finished it yet, and I expect it to take another 3 months or so..

**Comment by fayeed** [comment]
@cmpute No worries, I am still a beginner in Rust, Scryer & Dashu was the first proper Rust codebase I looked through missed a lot of things, anyway Thanks for the help with this.

**Comment by cmpute** [comment]
> @cmpute How do I compare `RBig` with `IBig` looks like only `IBig` implements `NumOrd`, thanks.

Ah yes, I'm working on implementing that between Ubig/IBig and Rbig

**Comment by cmpute** [comment]
The latest commit on the `dashu` repo now contains the implementation of `NumOrd<IBig>` for `RBig`.

**Comment by cmpute** [comment]
I have completed all the break changes I would like to include in the next version (v0.4) of `dashu` and it's ready for release. You can compose the PR based on the latest commit on the master branch. I would like to coordinate with you and see if you want to see more features to the next release of `dashu`. Thanks.

**Comment by fayeed** [comment]
Hey, I am PR open with the master on Scryer now, was able to get everything working, if you can release a new version I could update the PR with it, thanks.

**Comment by cmpute** [comment]
I have released the v0.4 of `dashu` crates. Thanks for your work on Scryer side!

**Pull request closed without merging**